lib/AzureMetadata/AzureDisk.pm

```diff
@@ -17,20 +17,55 @@ package AzureDisk;
 use strict;
 use warnings;
 
+use File::Basename;
 use POSIX;
 
+
 our @ISA    = qw (Exporter);
 our @EXPORT_OK = qw (
     getVHDDiskTag
 );
 
+sub get_root_device {
```

Comment:
looks pretty complex to me. Any reason why:

```
findmnt -n -f -o SOURCE /
```

does not work for you ?

Reply:
Other than that I would care for:


Subroutine "get_root_device" does not end with "return" at line 29, column 1.  See page 197 of PBP.  (Severity: 4)

Return value of "open" ignored at line 40, column 13.  Check the return value of "open" for success.  (Severity: 3)

Reply:
Was trying to avoid shelling out.

Reply:
@ikapelyukhin yes, the python implementation needs to find the root device, again preferably without shelling out

Reply:
what's wrong on shelling out ? I think the code here could result in some issues that will also be hard to debug

a) $block_device_name = basename($block_device)

    You assume the name of the block device symlink in /sys/block is also used in the name of the
    partitions on that block device. This is an assumption that is not necessarily correct, especially
    on devices that are mapping, like dm, lvm and friends

b) Next you raw read any partition device to get the major and minor and compare that with the
    stat result from /. Why not again call stat on the device to get that information ? It looks relatively complex to me.

I mean the entire block of 28lines of code with no test could be replaced by one findmnt. That should be worth a thought, shouldn't it ?

Thanks

Reply:
Really don't understand why even the python code should not shell out. findmnt belongs to util-linux, installed everywhere and required for any OS to function (provides e.g mount). Please can we stop re-inventing the wheel
Doesn't make any sense to me